crates/starknet-types-core/src/felt/num_traits_impl.rs

```diff
@@ -121,6 +123,39 @@ impl Pow<u128> for Felt {
     }
 }
 
+impl CheckedAdd for Felt {
+    fn checked_add(&self, v: &Self) -> Option<Self> {
+        let res = self + v;
+        if res < *self {
+            None
+        } else {
+            Some(res)
+        }
+    }
+}
+
+impl CheckedMul for Felt {
+    fn checked_mul(&self, v: &Self) -> Option<Self> {
+        let res = self * v;
+        if res < *self {
```

Comment:
I believe that implementation is wrong,
Felt::TWO * Felt::MAX will probably not work the same way as `2u32.wrapping_add(u32::MAX)` returns 1

Reply:
You are right. It should check against `max(self, v)`

I fixed it and added a test

Reply:
but now there is a bug for mul by zero

Reply:
fixed